File: javascript/secret-handshake/secret_handshake.js
var map = ["wink", "double blink", "close your eyes", "jump"];

function Handshake(number){
  if(typeof number != "number") throw 'Handshake must be a number';
  this.decimal = number;
}

Handshake.prototype.commands = function(){
  var binaryArray = this.decimal.toString(2).split("").reverse();
  var result = map.filter(function(e,i){ return binaryArray[i]==="1";});
  if(binaryArray.length > map.length) result.reverse();
  return result;
}

module.exports = Handshake;

File: go/secret-handshake/secret_handshake.go
package secret

const testVersion = 2

var actions = [4]string{"wink", "double blink", "close your eyes", "jump"}

// Handshake determines the handshake signals from the input number
func Handshake(number uint) []string {
	result := make([]string, 0)
	for i, action := range actions {
		if number&(uint(1)<<uint(i)) > 0 {
			result = append(result, action)
		}
	}
	if number&16 > 0 {
		reverse(result)
	}
	return result
}

// reverse reverses the order of the strings in an array in place
func reverse(stringList []string) {
	last := len(stringList) - 1
	for i := 0; i < len(stringList)/2; i++ {
		stringList[i], stringList[last-i] = stringList[last-i], stringList[i]
	}
}
